### PR TITLE
feature: Allow custom output for RepackModelStep

### PR DESCRIPTION
**Pull request by tuliocasagrande**
*Issue #, if available:*
#2803

*Description of changes:*
As proposed in #2803:
1. Add new parameter `repack_output_path` (default `None`) to `workflow.step_collections.RegisterModel`, `workflow.step_collections.EstimatorTransformer` and `workflow._utils._RepackModelStep`.
2. On internal `_RepackModelStep()` calls, handover `repack_output_path=repack_output_path` received from public `RegisterModel()` and `EstimatorTransformer()`
3. Inside of `_RepackModelStep`, send `output_path=repack_output_path` and `code_location=repack_output_path` to the internal `SKLearn()` call.

*Testing done:*

Tested RegisterModel without `repack_output_path` (current behavior):

```python
    step_register = RegisterModel(
        name="RegisterModel",
        model=pipeline_model,
        content_types=["text/csv"],
        response_types=["text/csv"],
        inference_instances=["ml.t2.medium", "ml.m5.large"],
        transform_instances=["ml.m5.large"],
        model_package_group_name=model_package_group_name,
        approval_status=model_approval_status,
        model_metrics=model_metrics,
    )
```

Current behaviour stayed the same on the generated pipeline:

```json
...
"OutputDataConfig": {
    "S3OutputPath": "s3://<my-default-bucket>/"
},
...
```

Tested RegisterModel with `repack_output_path`:

```python
    step_register = RegisterModel(
        name="RegisterModel",
        model=pipeline_model,
        content_types=["text/csv"],
        response_types=["text/csv"],
        inference_instances=["ml.t2.medium", "ml.m5.large"],
        transform_instances=["ml.m5.large"],
        model_package_group_name=model_package_group_name,
        approval_status=model_approval_status,
        model_metrics=model_metrics,
        repack_output_path=f"s3://{default_bucket}/{base_job_prefix}/repack",
    )
```
New output on RepackModel step:
```json
...
"OutputDataConfig": {
    "S3OutputPath": "s3://<my-default-bucket>/<my-base-job-prefix>/repack"
},
...
```

Added unit test `tests/unit/sagemaker/workflow/test_utils.py::test_repack_model_step_with_output_path`.

## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

#### General

- [X] I have read the [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md) doc
- [x] I certify that the changes I am introducing will be backword compatible, and I have discussed concerns about this, if any, with the Python SDK team
- [X] I used the commit message format described in [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md#committing-your-change)
- [X] I have passed the region in to all S3 and STS clients that I've initialized as part of this change.
- [x] I have updated any necessary documentation, including [READMEs](https://github.com/aws/sagemaker-python-sdk/blob/master/README.rst) and [API docs](https://github.com/aws/sagemaker-python-sdk/tree/master/doc) (if appropriate)

#### Tests

- [x] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [x] I have added unit and/or integration tests as appropriate to ensure backward compatibility of the changes
- [x] I have checked that my tests are not configured for a specific region or account (if appropriate)
- [x] I have used [`unique_name_from_base`](https://github.com/aws/sagemaker-python-sdk/blob/master/src/sagemaker/utils.py#L77) to create resource names in integ tests (if appropriate)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by tuliocasagrande** [comment]
@staubhp I finished this proposal.
I'd love to hear any feedback when you have the chance. Thanks for reviewing!

**Comment by staubhp** [comment]
Looks good to me, thanks for the contribution. Would you mind adding unit tests for RegisterModel & EstimatorTransformer step [here](https://github.com/aws/sagemaker-python-sdk/blob/dev/tests/unit/sagemaker/workflow/test_step_collections.py) just to confirm the repack_path is being wired through to _RepackModelStep?

**Comment by tuliocasagrande** [comment]
hey @staubhp, it was a great suggestion to add unit tests for RegisterModel & EstimatorTransformer.

I eventually discovered an issue with the `entry_point` parameter on `EstimatorTransformer` being propagated until `Model` and generating an error. We also didn't have unit tests for EstimatorTransformer with _RepackModelStep.

So I split the new changes in two commits:
1. Removed entry_point from kwargs after being used (similar to RegisterModel) and added unit test for this case: ee6afcf65fb9ea7cf771fe0b277782bfe16ffb01
2. Added unit tests for RegisterModel & EstimatorTransformer with custom repack output path: faf4ad585452e68780157fa108ba93573b4d5e1f

Please let me know what you think. Thanks for reviewing!

**Comment by staubhp** [comment]
Looks good to me. Thanks for the contribution @tuliocasagrande!

**Review by staubhp — APPROVED** [review]

File: src/sagemaker/workflow/step_collections.py [hunk 1]
```diff
@@ -336,6 +349,8 @@ def __init__(
                 security_group_ids=estimator.security_group_ids,
                 description=description,
                 display_name=display_name,
+                repack_output_path=repack_output_path,
```
Comment (eugeneteoh) [inline src/sagemaker/workflow/step_collections.py:352]: Can you add a `repack_image_uri` argument? My customer is facing an issue where the python version `SKLearn` (3.7) estimator is different from the training image uri. That caused dependency errors when trying to install packages that are available on (for example) Python 3.8.

File: src/sagemaker/workflow/step_collections.py [hunk 2]
```diff
@@ -317,16 +327,19 @@ def __init__(
...
+        repack_model = False
+
         if "entry_point" in kwargs:
-            entry_point = kwargs["entry_point"]
-            source_dir = kwargs.get("source_dir")
-            dependencies = kwargs.get("dependencies")
+            repack_model = True
+            entry_point = kwargs.pop("entry_point", None)
+            source_dir = kwargs.pop("source_dir", None)
+            dependencies = kwargs.pop("dependencies", None)
```
Comment (eugeneteoh) [inline src/sagemaker/workflow/step_collections.py:336]:
```suggestion
            dependencies = kwargs.pop("dependencies", None)
            code_location = kwargs.pop("code_location", None)
```

File: src/sagemaker/workflow/step_collections.py [hunk 3]
```diff
@@ -317,16 +327,19 @@ def __init__(
...
             repack_model_step = _RepackModelStep(
                 name=f"{name}RepackModel",
                 depends_on=depends_on,
                 retry_policies=repack_model_step_retry_policies,
                 sagemaker_session=estimator.sagemaker_session,
-                role=estimator.sagemaker_session,
+                role=estimator.role,
                 model_data=model_data,
                 entry_point=entry_point,
                 source_dir=source_dir,
```
Comment (eugeneteoh) [inline src/sagemaker/workflow/step_collections.py:345]:
```suggestion
                source_dir=source_dir,
                code_location=code_location,
```

Comment (eugeneteoh) [inline src/sagemaker/workflow/step_collections.py:345]:
```suggestion
                source_dir=source_dir,
                code_location=code_location,
```

**Review by eugeneteoh — CHANGES_REQUESTED** [review]
https://github.com/aws/sagemaker-python-sdk/pull/2804/files#diff-043e8c79b0b9ccd41f8164a29ccc1f9e67f38e349aafffb96cd91762c3824bc7R364
I suggest changing this to `image_uri=image_uri`. Using training_image_uri() for inference doesn't actually work.

File: src/sagemaker/workflow/step_collections.py (hunk 1 above)
Comment (eugeneteoh) [inline src/sagemaker/workflow/step_collections.py:352]: Can you add a `repack_image_uri` argument? My customer is facing an issue where the python version `SKLearn` (3.7) estimator is different from the training image uri. That caused dependency errors when trying to install packages that are available on (for example) Python 3.8. Another option would be equating it to `estimator.training_image_uri()`.
```suggestion
                repack_output_path=repack_output_path,
                image_uri=estimator.training_image_uri(),
```

**Review by eugeneteoh — CHANGES_REQUESTED** [review]

**Comment by qidewenwhen** [comment]
Update: this recently released [commit](https://github.com/aws/sagemaker-python-sdk/commit/cdd8cb5335ab0bdc772445d66dc38070209ba3a6) has made the repack step output path align with model repack path. In other words, users can use the `code_location` argument in model class to customize the output for `RepackModelStep`.

**Comment by tuliocasagrande** [comment]
Agree. Superseded by #3257.

**Pull request closed without merging**